### DISCUSSION TITLE
Array of Data objects serialisation fails for array properties

### DISCUSSION THREAD

**Discussion by ana28p** [spatie/laravel-data#826] [Q&A] [answered]
Hello,

I'm having issue with the data serialization after upgrading to version 4. In the controller I'm using `response()->json($myArrayOfDataObjects)` to get a `JsonResponse` and I'm getting the following error:

```
Spatie\LaravelData\Resolvers\TransformedDataCollectableResolver::Spatie\LaravelData\Resolvers\{closure}(): Argument #1 ($data) must be of type Spatie\LaravelData\Contracts\BaseData, array given, called in /var/www/html/vendor/spatie/laravel-data/src/Resolvers/TransformedDataCollectableResolver.php on line 80
```

I tried `response()->json()` on each property of one of the data objects in the array and it works fine, but not for one entire data object.

The issue might be that at some point a property of the data object is set as array from a collection of other data objects (e.g. `collect($childrenDataObjects)->toArray()`) which seems to transform also the data objects to arrays (so instead of having an array of data objects, I get an array of arrays).

Here is a small replication of the issue (using `Pest` test):
```
class Child extends \Spatie\LaravelData\Data
{
    public function __construct(
        public string $name,
    ) {}
}
class MainObject extends \Spatie\LaravelData\Data
{
    public function __construct(
        public string $name,
        /** @var array<Child> $children */
        public array $children,
    ) {}
}

it('data', function () {
    $children = collect(['child 1', 'child 2'])
        ->map(fn ($name) => new Child($name))
        ->toArray();
    // or
//    $children = Collection::make([
//        new Child('child 1'),
//        new Child('child 2'),
//    ])->toArray();
    $mainObject = new MainObject('parent', $children);

    $asJsonArray = $mainObject->jsonSerialize(); // fails
    expect($asJsonArray)->toBeArray();
    $asJsonResponse = response()->json($mainObject); // fails
    expect($asJsonResponse)->toBeInstanceOf(\Illuminate\Http\JsonResponse::class);
});
```
Laravel Data: 4.7.1
PHP: 8.3

It passes with
```
    $children = Child::collect([
        new Child('child 1'),
        new Child('child 2'),
    ]);
```

With v3, this works fine.
FIY: in my implementation the array of children is created from a `collection->map->toArray`

Is this a bug or is the behaviour intended? 
Are there any suggestions for a fix? Maybe something to make sure that `toArray()` won't transform the Data objects as well to arrays.

**Comment by rubenvanassche** [comment, marked as answer]
Hi @ana28p 

This is expected behavior, calling `toArray` transforms everything into an array even data objects (that's how Laravel works). You better use `all()` in such cases.

**Reply by ana28p** [reply]
Hi @rubenvanassche,

`all()` is indeed what I need. I missed that function. Thank you!
